### PR TITLE
Binary group fairness metrics

### PR DESCRIPTION
**Pull request by AndresAlgaba**
## What does this PR do?
This PR initializes the addition of [observational group fairness metrics](https://www.nature.com/articles/s41598-022-07939-1). The general idea of these metrics is to compare the model's outputs across different groups created by the protected attribute(s) under evaluation. 

As a first step, I implemented two common group fairness metrics, demographic parity and equal opportunity, for binary classification problems. For demographic parity, we compare the positivity rates and, for equal opportunity, the true positive rates. In the case of more than two groups, we use the largest disparity, i.e., dividing the lowest rate between the highest. 

In this initial proposal, I build on the ``stat_scores`` and add some logic to compute the positivity and true positive rates for groups. There are still several issues, but I believe that this proposal allows for a more clear discussion. For example, I'm unsure whether the ``_binary_groups_stat_scores`` should be part of ``stat_scores`` or ``group_fairness``. Overall, I tried to follow the style and logic of the other classification metrics as closely as possible.

I will also add documentation and appropriate testing at a later stage when the big decisions on the design of the API for group fairness metrics are settled. There are also several potential extensions, such as including additional classification metrics for detecting discrimination and adding a similar logic to regression-based metrics.

I leave it as a draft for now. @Borda 

## Before submitting
- [ ] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: src/torchmetrics/classification/__init__.py [hunk 1]
```diff
@@ -55,6 +55,8 @@
     MultilabelF1Score,
     MultilabelFBetaScore,
 )
+
+# from torchmetrics.classification.group_fairness import BinaryFairness, BinaryGroupStatRates
```
Comment (AndresAlgaba) [inline src/torchmetrics/classification/__init__.py:59]: I get the following error when adding these to the init file:

``` bash
python cannot import name 'metric' from partially initialized module 'torchmetrics' 
(most likely due to a circular import)
```

File: src/torchmetrics/classification/group_fairness.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+        group_stats = _binary_groups_stat_scores(
+            preds, target, groups, self.num_groups, self.threshold, self.ignore_index, self.validate_args
+        )
+
+        self._update_states(group_stats)
+
+    def compute(
+        self,
+    ) -> Union[Dict[str, torch.Tensor], Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]]:
```
Comment (AndresAlgaba) [inline src/torchmetrics/classification/group_fairness.py:246]: I'm unsure whether a ``Tuple``is the right choice when more than 3-4 group-fairness metrics are added. Maybe I should change this to a dict of dicts?

**Comment by Borda** [comment]
hi @stancld or @lucadiliello could you pls assist here with this PR?

File: src/torchmetrics/functional/classification/group_fairness.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+
+    group_preds = [preds[groups == i].unsqueeze(1) for i in range(num_groups)]
+    group_target = [target[groups == i].unsqueeze(1) for i in range(num_groups)]
+
+    group_stats = {
+        f"group_{group}": _binary_stat_scores_update(group_preds.pop(0), group_target.pop(0))
+        for group in range(num_groups)
+    }
+
+    return group_stats
```
Comment (lucadiliello) [inline src/torchmetrics/functional/classification/group_fairness.py:82]: I do not get why `_binary_groups_stat_scores` returns a dictionary and why `_update_states` accepts a dictionary since the keys seems not to be used. I think `_binary_groups_stat_scores` could just return a nested list and `_update_states` could just iterate over it for the internal updates.

File: src/torchmetrics/functional/classification/group_fairness.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+        _binary_stat_scores_arg_validation(threshold, "global", ignore_index)
+        _binary_stat_scores_tensor_validation(preds, target, "global", ignore_index)
+        _groups_validation(groups, num_groups)
+
+    preds, target = _binary_stat_scores_format(preds, target, threshold, ignore_index)
+    groups = _groups_format(groups)
+
+    group_preds = [preds[groups == i].unsqueeze(1) for i in range(num_groups)]
+    group_target = [target[groups == i].unsqueeze(1) for i in range(num_groups)]
+
```
Comment (lucadiliello) [inline src/torchmetrics/functional/classification/group_fairness.py:76]: This looks inefficient because you are indexing many times the same vector. I suggest taking a look at the retrieval metrics, in which a similar problem is solved with a single indexing.
Take a look here: https://github.com/Lightning-AI/metrics/blob/d2ef383c16f52f02ff38e7c7b8146010e694036b/src/torchmetrics/retrieval/base.py#L119

File: src/torchmetrics/functional/classification/group_fairness.py [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+    group_stats = {
+        f"group_{group}": _binary_stat_scores_update(group_preds.pop(0), group_target.pop(0))
+        for group in range(num_groups)
+    }
+
```
Comment (lucadiliello) [inline src/torchmetrics/functional/classification/group_fairness.py:81]: Pop in position 0 is expensive, why not something like:

```python
group_stats = {
         f"group_{group}": _binary_stat_scores_update(group_p, group_t)
         for group, (group_p, group_t) in enumerate(zip(group_preds, group_target))
     }
```

**Review by lucadiliello — COMMENTED** [review]
Hello, thanks for your contribution! I added some suggestions to improve code speed. Since I solved a similar problem when grouping metrics by indexes, if you need help feel free to ask!

File: src/torchmetrics/functional/classification/group_fairness.py (hunk 3 above)
Comment (AndresAlgaba) [reply src/torchmetrics/functional/classification/group_fairness.py:86]: You're right! I changed the ``Dict[str, Tuple[...]]`` to ``List[Tuple[...]]``.

File: src/torchmetrics/functional/classification/group_fairness.py (hunk 4 above)
Comment (AndresAlgaba) [reply src/torchmetrics/functional/classification/group_fairness.py:78]: Thanks for the example! I believe I've used a similar technique now to use single indexing.

**Comment by AndresAlgaba** [comment]
I have added documentation.

**Comment by AndresAlgaba** [comment]
Would adding other group fairness metrics for classification in this PR make sense? Or extending it to multi-class classification? Or should I keep this for a separate PR?

File: src/torchmetrics/classification/__init__.py (hunk 1 above)
Comment (stancld) [reply src/torchmetrics/classification/__init__.py:59]: Hi @AndresAlgaba, you should import `Metric` as `from torchmetrics.metric import Metric` instead of leveraging `torchmetrics import Metric`. This change should resolve the circular import issue.

Comment (AndresAlgaba) [reply src/torchmetrics/classification/__init__.py]: Thanks for the help! This indeed resolved the issue :)

**Comment by AndresAlgaba** [comment]
I will work on the testing and mark this PR ready for review.

File: tests/unittests/classification/inputs.py [hunk 6]
```diff
@@ -32,6 +32,7 @@ def _logsoftmax(x: Tensor, dim: int = -1) -> Tensor:
 
 
 Input = namedtuple("Input", ["preds", "target"])
+GroupInput = namedtuple("GroupInput", ["preds", "target", "groups"])
```
Comment (AndresAlgaba) [inline tests/unittests/classification/inputs.py:35]: I have added group inputs as a "random" tensor with groups is also required.

File: tests/unittests/helpers/testers.py [hunk 7]
```diff
@@ -663,3 +663,13 @@ def remove_ignore_index(target: Tensor, preds: Tensor, ignore_index: Optional[in
         idx = target == ignore_index
         target, preds = deepcopy(target[~idx]), deepcopy(preds[~idx])
     return target, preds
+
+
+def remove_ignore_index_groups(
```
Comment (AndresAlgaba) [inline tests/unittests/helpers/testers.py:668]: I am not sure whether I should create a separate function for this, or just modify the existing ``remove_ignore_index`` to accommodate groups?

File: tests/unittests/classification/test_group_fairness.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+from torchmetrics.classification.group_fairness import BinaryFairness
+from torchmetrics.functional.classification.group_fairness import binary_fairness
+from unittests.classification.inputs import _group_cases
+from unittests.helpers import seed_all
+from unittests.helpers.testers import THRESHOLD, MetricTester, inject_ignore_index, remove_ignore_index_groups
+
+seed_all(42)
+
+
+def _fairlearn_binary(preds, target, groups, ignore_index):
```
Comment (AndresAlgaba) [inline tests/unittests/classification/test_group_fairness.py:31]: This is only a draft version of the testing as I get slightly different results when computing the true positives rates in the fairlearn library (see: https://github.com/fairlearn/fairlearn/issues/1173).

File: tests/unittests/classification/test_group_fairness.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            target = inject_ignore_index(target, ignore_index)
+
+        self.run_class_metric_test(
+            ddp=ddp,
+            preds=preds,
+            target=target,
+            metric_class=BinaryFairness,
+            reference_metric=partial(_fairlearn_binary, ignore_index=ignore_index),
+            metric_args={"threshold": THRESHOLD, "ignore_index": ignore_index, "num_groups": 2, "task": "all"},
+            groups=groups,
```
Comment (AndresAlgaba) [inline tests/unittests/classification/test_group_fairness.py:66]: By providing the groups tensor as a ``kwarg`` I am afraid that it will not always be on the right device, but accounting for this seems to require some big changes. Therefore, I will mark this draft as ready for review to get some feedback on this :).

File: tests/unittests/classification/test_group_fairness.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+            preds = sigmoid(preds)
+        preds = (preds >= THRESHOLD).astype(np.uint8)
+
+    target, preds, groups = remove_ignore_index_groups(target, preds, groups, ignore_index)
+
+    dp = demographic_parity_ratio(target, preds, sensitive_features=groups)
+    eo = true_positive_rate_ratio(target, preds, sensitive_features=groups)
+
+    # TODO: need to create a Dict here
+    return dp, eo
```
Comment (AndresAlgaba) [inline tests/unittests/classification/test_group_fairness.py:47]: I believe the tests will certainly fail because I need to output a ``Dict`` here. I need to figure out how to get the appropriate keys.

File: tests/unittests/classification/test_group_fairness.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+            preds = sigmoid(preds)
+        preds = (preds >= THRESHOLD).astype(np.uint8)
+
+    target, preds, groups = remove_ignore_index_groups(target, preds, groups, ignore_index)
+
+    dp = demographic_parity_ratio(target, preds, sensitive_features=groups)
+    eo = true_positive_rate_ratio(target, preds, sensitive_features=groups)
+
+    # TODO: need to add groups here
+    return {"DP": dp, "EO": eo}
```
Comment (AndresAlgaba) [inline tests/unittests/classification/test_group_fairness.py:47]: The fairlearn package (and, by extension, most packages that compute group fairness metrics) return a ``float`` with the metric without indicating which groups have a min and max score. If we agree that it would be a good addition, I will write some "custom" code to figure out the min and max scores to add them to the dict keys.

File: tests/unittests/classification/test_group_fairness.py (hunk 9 above)
Comment (AndresAlgaba) [reply tests/unittests/classification/test_group_fairness.py:72]: Update: I am unsure how to proceed with the testing as adding them as ``kwargs_update: Additional keyword arguments that will be passed with preds and target when running update on the metric`` seems to cause many problems.

File: tests/unittests/classification/test_group_fairness.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+    mf = MetricFrame(metrics=metrics, y_true=target, y_pred=preds, sensitive_features=groups)
+
+    mf_group = mf.by_group
+    ratios = mf.ratio()
+
+    return {
+        f"DP_{mf_group['dp'].idxmin()}_{mf_group['dp'].idxmax()}": ratios["dp"],
+        f"EO_{mf_group['eo'].idxmin()}_{mf_group['eo'].idxmax()}": ratios["eo"],
+    }
```
[the remark this answers is not in the document]
Comment (Borda) [reply tests/unittests/classification/test_group_fairness.py:53]: I would unify it... :)

File: tests/unittests/helpers/testers.py (hunk 7 above)
Comment (SkafteNicki) [reply tests/unittests/helpers/testers.py:667]: sounds good to me :)

File: src/torchmetrics/functional/classification/group_fairness.py [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+def _groups_stat_transform(
+    group_stats: List[Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]]
+) -> Dict[str, torch.Tensor]:
+    """Transform group statistics by creating a tensor for each statistic."""
+    return {
+        "tp": torch.cat([stat[0].unsqueeze(0) for stat in group_stats]),
+        "fp": torch.cat([stat[1].unsqueeze(0) for stat in group_stats]),
+        "tn": torch.cat([stat[2].unsqueeze(0) for stat in group_stats]),
+        "fn": torch.cat([stat[3].unsqueeze(0) for stat in group_stats]),
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/classification/group_fairness.py:104]: pretty sure that stack = unsqueeze + cat
```suggestion
        "tp": torch.stack([stat[0] for stat in group_stats]),
        "fp": torch.stack([stat[1] for stat in group_stats]),
        "tn": torch.stack([stat[2] for stat in group_stats]),
        "fn": torch.stack([stat[3] for stat in group_stats]),
```

File: src/torchmetrics/classification/group_fairness.py [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+
+    Returns:
+        The metric returns a dict with a group identifier as key and a tensor with the tp, fp, tn and fn rates as value.
+
+    Example (preds is int tensor):
+        >>> from torchmetrics.classification import BinaryGroupStatRates
+        >>> target = torch.tensor([0, 1, 0, 1, 0, 1])
+        >>> preds = torch.tensor([0, 1, 0, 1, 0, 1])
+        >>> groups = torch.tensor([0, 1, 0, 1, 0, 1])
+        >>> metric = BinaryGroupStatRates(2)
```
[the remark this answers is not in the document]
Comment (AndresAlgaba) [reply src/torchmetrics/classification/group_fairness.py:78]: Thanks @SkafteNicki for all the suggestions and feedback!

File: src/torchmetrics/classification/group_fairness.py [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+        self,
+    ) -> Dict[str, torch.Tensor]:
+        """Computes tp, fp, tn and fn rates based on inputs passed in to ``update`` previously."""
+        results = torch.stack((self.tp, self.fp, self.tn, self.fn), dim=1)
+
+        return {f"group_{i}": group / group.sum() for i, group in enumerate(results)}
+
+
+class BinaryFairness(_AbstractGroupStatScores):
+    r"""Computes demographic parity and equal opportunity ratio for binary classification problems.
```
[the remark this answers is not in the document]
Comment (AndresAlgaba) [reply src/torchmetrics/classification/group_fairness.py:141]: Sure! @aniketmaurya and I will be working on a blog post regarding this PR. Can we include that one later on?

File: src/torchmetrics/classification/group_fairness.py [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+    _binary_groups_stat_scores,
+    _compute_binary_demographic_parity,
+    _compute_binary_equal_opportunity,
+)
+from torchmetrics.functional.classification.stat_scores import _binary_stat_scores_arg_validation
+from torchmetrics.metric import Metric
+from torchmetrics.utilities import rank_zero_warn
+
+
+class _AbstractGroupStatScores(Metric):
```
[the remark this answers is not in the document]
Comment (AndresAlgaba) [reply src/torchmetrics/classification/group_fairness.py:29]: Hi @Borda , do you mean like an ABC? I was following the template from: https://github.com/Lightning-AI/metrics/blob/b5d8e7b73e19f358a168f958f32b04f7b13b35da/src/torchmetrics/classification/stat_scores.py#L41

[the remark this answers is not in the document]
Comment (justusschock) [reply src/torchmetrics/classification/group_fairness.py:29]: @Borda this is fine since it is abstract unless it implements `compute` and `update`

File: src/torchmetrics/classification/group_fairness.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply src/torchmetrics/classification/group_fairness.py:141]: that's definitely possible but it would be nice to have before this PR gets merged :)

File: tests/unittests/classification/test_group_fairness.py (hunk 9 above)
Comment (justusschock) [reply tests/unittests/classification/test_group_fairness.py:72]: I agree. I think you could modify the baseclass tester here to use https://github.com/Lightning-AI/utilities/blob/main/src/lightning_utilities/core/apply_func.py#L23 on all kwargs to move tensors to the correct devices (e.g. of `preds`)

**Review by justusschock — APPROVED** [review]

File: src/torchmetrics/classification/group_fairness.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (AndresAlgaba) [reply src/torchmetrics/classification/group_fairness.py:141]: Sure, working on it!

**Comment by AndresAlgaba** [comment]
@justusschock, thanks for the feedback! I've looked into the device setting and it is actually already implemented this way. The error occurred since I put the ``groups``  argument before the ``target`` argument, but fixed that now!

**Comment by AndresAlgaba** [comment]
Currently a problem remains in: 

https://github.com/Lightning-AI/metrics/blob/a754eb7b59850c9748fdd856188445e51780923d/tests/unittests/helpers/testers.py#L262

More specifically in:

https://github.com/Lightning-AI/metrics/blob/a754eb7b59850c9748fdd856188445e51780923d/tests/unittests/helpers/testers.py#L314

where 

https://github.com/Lightning-AI/metrics/blob/a754eb7b59850c9748fdd856188445e51780923d/tests/unittests/helpers/testers.py#L64

I get an error in 
```python
    elif isinstance(pl_result, Dict):
        if key is None:
            raise KeyError("Provide Key for Dict based metric results.")
        assert np.allclose(pl_result[key].detach().cpu().numpy(), sk_result, atol=atol, equal_nan=True)
```
since I return a ``Dict``, but do not provide any keys. Should I allow here for keys via ``kwargs``?

File: src/torchmetrics/classification/group_fairness.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Borda) [reply src/torchmetrics/classification/group_fairness.py:141]: @AndresAlgaba ^^

File: tests/unittests/classification/test_group_fairness.py [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+            metric_args={
+                "threshold": THRESHOLD,
+                "ignore_index": ignore_index,
+                "task": "all",
+            },
+            groups=groups,
+            fragment_kwargs=True,
+        )
+
+    # def test_binary_fairness_differentiability(self, input):
```
Comment (Borda) [inline tests/unittests/classification/test_group_fairness.py:96]: Why this is commented, then remove it?

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
> Currently a problem remains

@lucadiliello @justusschock could you pls help here? :rabbit:

File: tests/unittests/classification/test_group_fairness.py (hunk 17 above)
Comment (AndresAlgaba) [reply tests/unittests/classification/test_group_fairness.py]: Sorry, was temporary since they were still failing. Fixed it now.

File: src/torchmetrics/classification/group_fairness.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (AndresAlgaba) [reply src/torchmetrics/classification/group_fairness.py]: I've added a link for each metric, as the blogpost is not finished yet! We can maybe change this later?

**Comment by stancld** [comment]
Some test suites except for a few look happy. May I ask you, @Borda and @justusschock, to re-review my proposed changes in the test file, please? 👼

**Review by stancld — APPROVED** [review]
@AndresAlgaba Thanks for the great work! Except for properly handling a bit of corner-case tests, it really lgtm.

**Comment by AndresAlgaba** [comment]
@stancld, thank you very much for the feedback and help :D

**Review by justusschock — APPROVED** [review]
Changes LGTM!

@stancld thank you very much for your help!

**Review by lucadiliello — APPROVED** [review]

**Comment by stancld** [comment]
@Borda @justusschock Do you have any clue why our tests failing with some old configurations? 🤔

**Comment by Borda** [comment]
> @Borda @justusschock Do you have any clue why our tests failing with some old configurations? thinking

I think I remember this one, you are passing somewhere none instead of expected int:
```
TypeError: '<' not supported between instances of 'int' and 'NoneType'
```
I would suggest recreating this env and debugging locally... to do you you can use:
https://github.com/Lightning-AI/metrics/blob/50388cfb167ba323fcc407a89d83f2aec0dfb171/.github/actions/pull-caches/action.yml#L33

File: requirements/classification_test.txt [hunk 18]
```diff
@@ -2,3 +2,4 @@
 #  in case you want to preserve/enforce restrictions on the latest compatible version, add "strict" as an in-line comment
 
 netcal >1.0.0, <=1.3.3  # calibration_error
+fairlearn # group_fairness
```
Comment (stancld) [inline requirements/classification_test.txt:5]: I think we should also add `pandas` here.

**Comment by stancld** [comment]
> > @Borda @justusschock Do you have any clue why our tests failing with some old configurations? thinking
> 
> I think I remember this one, you are passing somewhere none instead of expected int:
> 
> ```
> TypeError: '<' not supported between instances of 'int' and 'NoneType'
> ```
> 
> I would suggest recreating this env and debugging locally... to do you you can use:
> 
> https://github.com/Lightning-AI/metrics/blob/50388cfb167ba323fcc407a89d83f2aec0dfb171/.github/actions/pull-caches/action.yml#L33

Cool, thanks for the advice. I've setup the docker and will try to figure out what's going on there.

**Comment by AndresAlgaba** [comment]
> > > @Borda @justusschock Do you have any clue why our tests failing with some old configurations? thinking
> > 
> > 
> > I think I remember this one, you are passing somewhere none instead of expected int:
> > ```
> > TypeError: '<' not supported between instances of 'int' and 'NoneType'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I would suggest recreating this env and debugging locally... to do you you can use:
> > https://github.com/Lightning-AI/metrics/blob/50388cfb167ba323fcc407a89d83f2aec0dfb171/.github/actions/pull-caches/action.yml#L33
> 
> Cool, thanks for the advice. I've setup the docker and will try to figure out what's going on there.

Hey @stancld, doing the same currently :D Let me know if I can be of any help! It indeed seems to be some interplay with the fairlearn library.

**Comment by stancld** [comment]
@Borda`fairlearn` declares supports for python `3.8` and newer. What about skipping tests for `python==3.7` then?

cc: @AndresAlgaba

**Comment by stancld** [comment]
Hi @AndresAlgaba, would you please check why `make-docs` job is failing? 🙏 Otherwise, everything looks good now :]

**Comment by stancld** [comment]
@AndresAlgaba -- Sorry, actually found out one link for `InfoLM` seems to be broken.

**Comment by Borda** [comment]
@AndresAlgaba thank you, and apology that it took so long...

**Comment by AndresAlgaba** [comment]
@Borda, my pleasure! And thanks to the entire team for all the help :D
